src/tree/updater_quantile_hist.cc

```diff
@@ -89,39 +88,163 @@ bool QuantileHistMaker::UpdatePredictionCache(
   }
 }
 
-void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
-                                        const GHistIndexBlockMatrix& gmatb,
-                                        const ColumnMatrix& column_matrix,
-                                        HostDeviceVector<GradientPair>* gpair,
-                                        DMatrix* p_fmat,
-                                        RegTree* p_tree) {
-  double gstart = dmlc::GetTime();
-
-  int num_leaves = 0;
+void QuantileHistMaker::Builder::ExpandWithDepthWidth(
+  const GHistIndexMatrix &gmat,
+  const GHistIndexBlockMatrix &gmatb,
+  const ColumnMatrix &column_matrix,
+  DMatrix *p_fmat,
+  RegTree *p_tree,
+  const std::vector<GradientPair> &gpair_h) {
   unsigned timestamp = 0;
+  int num_leaves = 0;
 
-  double tstart;
-  double time_init_data = 0;
-  double time_init_new_node = 0;
-  double time_build_hist = 0;
-  double time_evaluate_split = 0;
-  double time_apply_split = 0;
+  tstart = dmlc::GetTime();
 
-  const std::vector<GradientPair>& gpair_h = gpair->ConstHostVector();
+  this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
+  time_init_data = dmlc::GetTime() - tstart;
 
-  spliteval_->Reset();
+  // FIXME(hcho3): this code is broken when param.num_roots > 1. Please fix it
+  CHECK_EQ(p_tree->param.num_roots, 1) <<
+    "tree_method=hist does not support multiple roots at this moment";
 
+  // in depth_wise growing, we feed loss_chg with 0.0 since it is not used anyway
+  qexpand_depth_wise.push_back(ExpandEntry(0, p_tree->GetDepth(0), 0.0, timestamp++));
+  ++num_leaves;
+  // get stats of node per depth
   tstart = dmlc::GetTime();
-  this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
-  time_init_data = dmlc::GetTime() - tstart;
+  for (int depth = 0; depth < param_.max_depth + 1; depth++) {
```

Comment:
The logic inside this for loop contains many different cases and will be difficult to read and maintain. Can we further simplify this? If I understand correctly most of this code is evaluating whether to actually build the histogram or defer construction of the histogram and use the subtraction trick.